modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -74,48 +172,111 @@ object GlobalL0Service {
+              pulled <- pullSnapshots(l0Peer, lastStoredOrdinal.next, lastOrdinal)
+            } yield pulled.asRight[LatestSnapshotTuple]
           }
         }.handleErrorWith { e =>
-          logger.warn(e)(s"Failure pulling global snapshots!") >>
+          logger.warn(e)("Failure pulling global snapshots") >>
             Applicative[F].pure(
-              List.empty[Hashed[GlobalIncrementalSnapshot]].asRight[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)]
+              List.empty[Hashed[GlobalIncrementalSnapshot]].asRight[LatestSnapshotTuple]
             )
         }
+
+      private def pullSnapshots(
+        l0Peer: L0Peer,
+        nextOrdinal: SnapshotOrdinal,
+        lastOrdinal: SnapshotOrdinal
+      ): F[List[Hashed[GlobalIncrementalSnapshot]]] = {
+        val lastOrdinalCap = lastOrdinal.value.value
+          .min(singlePullLimit.map(nextOrdinal.value.value + _.value).getOrElse(lastOrdinal.value.value))
+
+        val ordinals = (nextOrdinal.value.value to lastOrdinalCap).toList
+          .map(o => SnapshotOrdinal(NonNegLong.unsafeFrom(o)))
+
+        (ordinals, List.empty[Hashed[GlobalIncrementalSnapshot]]).tailRecM {
+          case (ordinal :: nextOrdinals, snapshots) =>
+            l0GlobalSnapshotClient
+              .get(ordinal)(l0Peer)
+              .flatMap(_.toHashedWithSignatureCheck.flatMap(_.liftTo[F]))
+              .map(s => (nextOrdinals, snapshots :+ s).asLeft[List[Hashed[GlobalIncrementalSnapshot]]])
+              .handleErrorWith { e =>
+                logger
+                  .warn(e)(s"Failure pulling snapshot with ordinal=${ordinal.show}")
+                  .map(_ => snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
+              }
+
+          case (Nil, snapshots) =>
+            Applicative[F].pure(snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
+        }
+      }
+
+      private case class MajoritySnapshotData(ordinal: SnapshotOrdinal, hash: Hash)
+
+      private def getLatestMajoritySnapshotData(peerIds: NonEmptyList[PeerId]): F[MajoritySnapshotData] = {
+        val maybeData =
+          for {
+            peers <- OptionT(getL0Peers(peerIds))
+            majorityOrdinal <- OptionT(getMajorityOrdinal(peers))
+            majorityHash <- OptionT(getMajorityHash(peers, majorityOrdinal))
+          } yield MajoritySnapshotData(majorityOrdinal, majorityHash)
+
+        maybeData.getOrElseF(NoMajoritySnapshotData.raiseError[F, MajoritySnapshotData])
+      }
+
+      private def getMajorityHash(peers: NonEmptyList[L0Peer], ordinal: SnapshotOrdinal): F[Option[Hash]] =
+        peers.toList
+          .parTraverseN(numConcurrentQueries)(p =>
+            attemptOption(l0GlobalSnapshotClient.getHash(ordinal).run(p), s"Unable to obtain hash for majority peer ${p.show}")
+          )
+          .map(_.flatMap(_.flatten))
+          .flatTap(hashes => logger.debug(s"** Majority Hashes ${hashes.show}"))
```

Comment:
You will log a lot of repeated information (repeated hashes) and also in long format (every hash takes 64 chars AFAIR). I would not log until the majority information is simplified somehow - I guess hashes counted and the shorter hash format. Maybe the pickMajority function could return the picked element and metadata as well - this metadata could be logged if needed.

Reply:
It's logging at the DEBUG level -- isn't the point of logging something as DEBUG is so we can verify something? My understanding is that production deploys should not be set below INFO.

Reply:
@TheMMaciek Would this be preferable?
```
scala> Hash("12345abcdefg67890").debugString
val res5: String = Hash[12345...67890]
```

Reply:
Or, I can write method `hash.truncated` which creates `TruncatedHash` and have that class define a `show` method